### PR TITLE
grpc : Dynamic plugin loading

### PR DESCRIPTION
**Pull request by Rjasuja**
Split proto files(one for each plugin), load plugins based on plugins.conf file

### CODE REVIEW

**Comment by Rjasuja** [comment]
For now conf file is kept simple with plugin name in each new line, other option is to have xml file and have xml parser included.

**Comment by julianoes** [comment]
> For now conf file is kept simple with plugin name in each new line, other option is to have xml file and have xml parser included.

We will have a xml parser included once the camera definition stuff is added that you could use.

File: grpc/python_client/CMakeLists.txt [hunk 1]
```diff
@@ -16,4 +16,33 @@ if(${PYTHONINTERP_FOUND})
...
+    )
+
+    add_custom_command(TARGET pydronecore
+        COMMAND ${PYTHON_EXECUTABLE}
+            -m grpc_tools.protoc
+            -I ${proto_dir}
+            --python_out=.
+            --grpc_python_out=.
+            ${proto_dir}/telemetry.proto
+        DEPENDS ${proto_dir}/telemetry.proto
```
Comment (julianoes) [inline grpc/python_client/CMakeLists.txt]: With cmake you can list all files in a directory and loop over them.

File: grpc/server/CMakeLists.txt [hunk 2]
```diff
@@ -27,10 +27,64 @@ add_custom_command(OUTPUT dronecore.pb.cc
...
 )
 
+add_custom_command(OUTPUT action.grpc.pb.cc
+    COMMAND ${Protobuf_PROTOC_EXECUTABLE}
+        -I ${proto_dir}
+        --grpc_out=.
+        --plugin=protoc-gen-grpc=${GRPC_CPP_PLUGIN}
+        --cpp_out=. ${proto_dir}/action.proto
+    DEPENDS ${proto_dir}/action.proto
+)
```
Comment (julianoes) [inline grpc/server/CMakeLists.txt]: Also, here I suggest to loop over the files automatically.

File: grpc/server/dronecore_server.cpp [hunk 3]
```diff
@@ -21,16 +28,34 @@ using grpc::ServerReaderWriter;
...
+
 class DroneCoreRPCImpl final : public DroneCoreRPC::Service
 {
 
 public:
-    Status Arm(ServerContext *context, const Empty *request,
+
+};
+
+class ActionRPCImpl final : public ActionRPC::Service
```
Comment (julianoes) [inline grpc/server/dronecore_server.cpp]: I think these classes should be in separate files and then inside the respective plugin folder.

File: grpc/proto/mission.proto [hunk 4]
```diff
@@ -0,0 +1,51 @@
+syntax = "proto3";
```
Comment (julianoes) [inline grpc/proto/mission.proto]: I think these proto files should be moved to the respective plugin folders.

File: grpc/server/dronecore_server.cpp [hunk 5]
```diff
@@ -111,25 +140,59 @@ class DroneCoreRPCImpl final : public DroneCoreRPC::Service
...
+
 void RunServer()
 {
     std::string server_address("0.0.0.0:50051");
     DroneCoreRPCImpl service;
 
+    map_type map;
+    map["action"] = &createInstances<ActionRPCImpl>;
+    map["mission"] = &createInstances<MissionRPCImpl>;
+    map["telemetry"] = &createInstances<TelemetryRPCImpl>;
```
Comment (julianoes) [inline grpc/server/dronecore_server.cpp]: This will need to be auto-generated in the same way as the plugins are currently added to device_impl.

**Comment by avinash-palleti** [comment]
@julianoes Can you review the changes. If this is fine, we can go ahead for multiple device handling implementation through gRPC.

File: grpc/server/CMakeLists.txt [hunk 6]
```diff
@@ -27,10 +29,39 @@ add_custom_command(OUTPUT dronecore.pb.cc
...
 add_executable(dronecore_server
     dronecore_server.cpp
     dronecore.pb.cc # generated
     dronecore.grpc.pb.cc # generated
+	action.pb.cc # generated
+    action.grpc.pb.cc # generated
+    mission.pb.cc # generated
+    mission.grpc.pb.cc # generated
+    telemetry.pb.cc # generated
+    telemetry.grpc.pb.cc # generated
```
Comment (julianoes) [inline grpc/server/CMakeLists.txt]: You could also generate that list using the same `foreach`. You just add it to all to a variable.

File: grpc/server/dronecore_server.cpp (hunk 5 above)
Comment (julianoes) [reply grpc/server/dronecore_server.cpp]: Do you agree?

File: grpc/server/plugins/action/action.proto [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+package dronecorerpc;
+
+service ActionRPC {
+    rpc Arm(ActionEmpty) returns(ActionResult) {}
+    rpc TakeOff(ActionEmpty) returns(ActionResult) {}
+    rpc Land(ActionEmpty) returns(ActionResult) {}
+}
+
+// FIXME these different Empty types are ugly
```
Comment (julianoes) [inline grpc/server/plugins/action/action.proto]: I think there is a protobuf empty that we should reuse: `import "google/protobuf/empty.proto";`.

File: grpc/server/plugins/telemetry/telemetry.proto [hunk 8]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
+
+package dronecorerpc;
+
+service TelemetryRPC {
+    rpc TelemetryPositionSubscription(TelemetryEmpty) returns(stream TelemetryPosition) {}
+}
+
+message TelemetryEmpty {}
```
Comment (julianoes) [inline grpc/server/plugins/telemetry/telemetry.proto]: Also re-use empty here.

File: grpc/server/dronecore_server.cpp (hunk 5 above)
Comment (avinash-palleti) [reply grpc/server/dronecore_server.cpp]: @julianoes createInstance takes class name as parameter, if not here somewhere there should be mapping between string to class name. Any other way we could do without hardcoded mapping.

Thanks for the review, we will address other comments in next submission.

**Comment by mrpollo** [comment]
Same comments as @julianoes , thanks for the PR @Rjasuja

File: grpc/server/plugins/action/actionrpc_impl.h [hunk 9]
```diff
@@ -1,22 +1,23 @@
 #include "action.h"
 #include "action.grpc.pb.h"
+#include <google/protobuf/empty.pb.h>
 
 using grpc::Status;
 using grpc::ServerContext;
 using dronecorerpc::ActionRPC;
-using dronecorerpc::ActionEmpty;
+using dronecorerpc::Empty;
```
Comment (julianoes) [inline grpc/server/plugins/action/actionrpc_impl.h]: Is this line still needed?

**Comment by Rjasuja** [comment]
Removed.

**Comment by Rjasuja** [comment]
Removed.

**Comment by Rjasuja** [comment]
@julianoes Added cmake file for automated generation of plugin objects. Please review.

**Comment by julianoes** [comment]
@Rjasuja cool thanks, this looks better! Can you merge or rebase with develop and check the CI issues please?

**Comment by julianoes** [comment]
Thanks @Rjasuja!

**Pull request merged**